doc/maybe.md

```diff
@@ -0,0 +1,74 @@
+# Maybe (template)
+
+Class `Napi::Maybe<T>` represents an maybe empty value: every `Maybe` is either
+`Just` and contains a value, or `Nothing`, and does not. `Maybe` types are very
+common in node-addon-api code, as they represent the function may throw a
+JavaScript exception and cause the program unable to evaluation any JavaScript
+code until the exception is been handled.
+
+Typically, the value wrapped in `Napi::Maybe<T>` is [`Napi::Value`] and its
+subclasses.
+
+## Methods
+
+### IsNothing
+
+```cpp
+template <typename T>
+bool Napi::Maybe::IsNothing() const;
+```
+
+Returns if the `Maybe` is `Nothing` and does not contain a value.
+
+### IsJust
+
+```cpp
+template <typename T>
+bool Napi::Maybe::IsJust() const;
+```
+
+Returns if the `Maybe` is `Just` and contains a value.
+
+### Check
+
+```cpp
+template <typename T>
+void Napi::Maybe::Check() const;
+```
+
+Short-hand for `Maybe::Unwrap()`, which doesn't return a value. Could be used
+where the actual value of the Maybe is not needed like `Object::Set`.
+If this Maybe is nothing (empty), node-addon-api will crash the
+process.
+
+### Unwrap
+
+```cpp
+template <typename T>
+T Napi::Maybe::Unwrap() const;
```

Comment:
I've chosen the three value unwrapping method to be named in a more intuitive pattern like `Unwrap()`, `UnwrapOr(const T& default_value)`, and `UnwrapTo(T* out)`. This is somewhat not aligned with v8's `Maybe` interface, which has `FromJust`, `FromMaybe`, `To` respectively in names.